parsing/extensions.ml

```diff
@@ -297,7 +297,36 @@ module Immutable_arrays = struct
 
   let of_pat expr = match expr.ppat_desc with
     | Ppat_array elts -> Iapat_immutable_array elts
-    | _ -> failwith "Malformed immutable array expression"
+    | _ -> failwith "Malformed immutable array pattern"
+end
+
+(** Strengthening *)
+module Strengthen = struct
+  type nonrec module_type =
+    { mty : Parsetree.module_type; mod_id : Longident.t Location.loc }
+
+  let extension_string = Language_extension.to_string Strengthen
+
+  (* Encoding: [S with M] becomes [functor (_ : S) -> M]. But remember
+     that this gets wrapped in an [[%extension.strengthen]] node, so
+     the what starts as [S with M] really becomes
+
+       {[
+         functor (_ : [%extension.strengthen]) ->
+           functor (_ : S) -> M
+       ]}
```

Comment:
I don't think this part of the comment belongs here; a user not reading `extensions_parsing.ml` should never have to think about the encoding we choose.  See the pseudo-BNF I have for comprehensions.